consensus/misc/eip4844.go

```diff
@@ -27,7 +30,59 @@ var (
 	dataGaspriceUpdateFraction = big.NewInt(params.BlobTxDataGaspriceUpdateFraction)
 )
 
+// VerifyEIP4844Header verifies the presence of the excessDataGas field and that
+// if the current block contains no transactions, the excessDataGas is updated
+// accordingly.
+//
+// We cannot verify excessDataGas if there *are* transactions included as that
+// would require the block body. However, it is nonetheless useful to verify the
+// header in case of an empty body since certain code might skip body validations
+// with no included transactions (e.g. snap sync).
+func VerifyEIP4844Header(parent, header *types.Header) error {
+	// Verify the header is not malformed
+	if header.ExcessDataGas == nil {
+		return errors.New("header is missing excessDataGas")
+	}
+	// Verify the excessDataGas is correct based on the parent header iff the
+	// transaction list is empty. For non-empty blocks, validation needs to be
+	// done later.
+	if header.TxHash == types.EmptyTxsHash {
```

Comment:
It seems pretty rare this code path would actually be triggered, what is benefit of fast failing here?

Reply:
Interestingly, the downloader has an optimisation that it only requests block bodies if they are not empty. If they are empty, then the bodies aren't requested and just pre-filled to zero.

Buuut, 4844's excessDataGas field *needs* the bodies to be verified. Since the downloader just blindly fills teh transaction list as empty in those cases, we have two options:

- Have the excessDataGas already validated like all other header fields (what this check does)
- Introduce header validation pathways into the "fill empty block" mechanism of the downloader

The second thing seemed wonkier as it would need changes that touch a number of unintuitive spots in the code.

But in all honesty, this change is also not that intuitive, it just seemed nicer.

Alternatively we could do something akin to the "gasUsed" which is completely ignored during snap sync and just assumed to be correct whatever it is, but here we not only have "blobsUsed" but also "pricing" so it seemed more relevant to try and validate it.

TL;DR: These are all fallouts of that convoluted field.